catalog/video/catalog-mpegts-stream/src/main/java/com/connexta/alliance/video/stream/mpegts/PacketBuffer.java

```diff
@@ -0,0 +1,405 @@
+
+        try (OutputStream os = outputStreamFactory.create(getTempFile(), true)) {
+
+            List<byte[]> outgoingPackets = frames.subList(0, index + 1)
+                    .stream()
+                    .flatMap(frame -> frame.packets.stream())
+                    .collect(Collectors.toList());
+            frames = new ArrayList<>(frames.subList(index + 1, frames.size()));
+
+            for (byte[] outgoingPacket : outgoingPackets) {
+                os.write(outgoingPacket);
+                bytesWrittenToTempFile += outgoingPacket.length;
+            }
+
+        }
+
+    }
+
+    /**
+     * If the rollover condition is not met, then the method will return {@link Optional#empty()}.
+     * If the rollover condition is met, then the method <b>may</b> return a temp file. The only
+     * reason a temp file may not be returned is when the condition is <code>true</code> even when
+     * no data has been written to file. The caller is responsible for deleting the temp file.
+     *
+     * @param rolloverCondition the rollover condition
+     * @return an optional temp file
+     */
+    public Optional<File> rotate(RolloverCondition rolloverCondition) {
+        lock.lock();
+        try {
+            if (!rolloverCondition.isRolloverReady(this)) {
+                return Optional.empty();
+            }
+            if (currentTempFile == null || bytesWrittenToTempFile == 0) {
+                return Optional.empty();
+            }
+            File tempFile = currentTempFile;
+            currentTempFile = null;
+            bytesWrittenToTempFile = 0;
+            return Optional.of(tempFile);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Flush all buffered data to disk and rotate. Will return {@link Optional#empty()} if no data
+     * has been written to file. The caller is responsible for deleting the temp file.
+     *
+     * @return an optional temp file
+     * @throws IOException
+     */
+    public Optional<File> flushAndRotate() throws IOException {
+        lock.lock();
+        try {
+
+            if (!incompleteFrame.isEmpty()) {
+                frames.add(new Frame(FrameType.UNKNOWN, incompleteFrame));
+                incompleteFrame = new ArrayList<>();
+            }
+
+            if (!frames.isEmpty()) {
+                flushFrameset(frames.size() - 1);
+            }
+
+            return rotate(ALWAYS_TRUE);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private File getTempFile() throws IOException {
+        if (currentTempFile == null) {
+            tempFileCreateTime = dateSupplier.get()
+                    .getTime();
+            bytesWrittenToTempFile = 0;
+            currentTempFile = tempFileGenerator.generate();
+        }
+        return currentTempFile;
+    }
+
+    private Map<FrameType, AtomicLong> summarizeFrameTypes() {
+
+        Map<FrameType, AtomicLong> frameTypeCounts = new HashMap<>();
```

Comment:
I could. I like using AtomicLong becuase it has an increment method, that way I don't have to get the value form the map and then put the new value.